### PR TITLE
Fix for duplicate events in AWS

### LINKED ISSUE #2585: Duplicated AWS Cloudtrail logs (3.8.0)
Hello team,

The AWS Woodle is duplicating logs in version 3.8.0 of Wazuh. The Woodle isn't ignoring the logs he's already read.

Regards,

Miguel Casares

### PR DESCRIPTION
**Pull request by druizz90**
Hi team,

This PR closes #2585, #2972 and #2982. I modified SQL queries for cleaning `AWS` database and I added a method to update `metadata` table.

Best regards,

Demetrio.

### CODE REVIEW

File: wodles/aws/aws-s3.py [hunk 1]
```diff
@@ -372,23 +372,34 @@ def __init__(self, reparse, access_key, secret_key, profile, iam_role_arn,
...
-                                    ROWID DESC
-                                    LIMIT {retain_db_records})"""
+                                    aws_region='{aws_region}';"""
 
         self.db_name = 's3_cloudtrail'
         WazuhIntegration.__init__(self, access_key=access_key, secret_key=secret_key,
             aws_profile=profile, iam_role_arn=iam_role_arn, bucket=bucket, service_name='s3')
         self.legacy_db_table_name = 'log_progress'
-        self.retain_db_records = 1000
+        self.retain_db_records = 500
```
[the remark this answers is not in the document]
Comment (druizz90) [reply wodles/aws/aws-s3.py]: 500 is a bigger enough for our purpose.

**Comment by UranusBytes** [comment]
@druizz90 

There is a typo in the aws-s3.py that is causing it to error out.

https://github.com/wazuh/wazuh/blob/12ab603460cbbb26344bf89de4e42ef1ceb87ed1/wodles/aws/aws_s3.py#L178

The line should be
`print('ERROR: Error creating metadata table: {e}'.format(e))`

Otherwise, FWIW I'm testing this version in my environment in place of the hack I'd implemented to resolve the problem.

Jeremy

**Comment by druizz90** [comment]
Hi @UranusBytes,

I am going to fix it, I forgot that formats strings are not available in Python 2 (unfortunately).

Our wodle will be continue compatible with Python 2 in the next releases.

Thank you for your review.

Best regards,

Demetrio.

**Comment by UranusBytes** [comment]
Hi @druizz90 ,

While the code does appear to have resolved duplicate events so far, on initial execution it was not able to continue with the last log processed, but instead resorted to the value defined for `only_logs_after` for the bucket in `ossec.conf`.  Before this, I had changed the value to 2019-MAR-31, so it only re-processed 3 days worth of events, but this could cause a significant reprocessing for many users.

Jeremy

**Comment by druizz90** [comment]
Hi @UranusBytes,

Could you give me more details about it? I can continue processing logs after last log which was processed properly.

Best regards,

Demetrio.

**Comment by druizz90** [comment]
## Tests performed

```bash
# /var/ossec/framework/python/bin/python3 -m pytest ./ -v
================================================ test session starts =================================================
platform linux -- Python 3.7.2, pytest-4.4.0, py-1.8.0, pluggy-0.9.0 -- /var/ossec/framework/python/bin/python3
cachedir: .pytest_cache
rootdir: /root/wazuh/wodles/aws
collected 7 items                                                                                                    

tests/test_aws.py::test_metadata_version_buckets[AWSCloudTrailBucket] PASSED                                   [ 14%]
tests/test_aws.py::test_metadata_version_buckets[AWSConfigBucket] PASSED                                       [ 28%]
tests/test_aws.py::test_metadata_version_buckets[AWSVPCFlowBucket] PASSED                                      [ 42%]
tests/test_aws.py::test_metadata_version_buckets[AWSCustomBucket] PASSED                                       [ 57%]
tests/test_aws.py::test_metadata_version_buckets[AWSGuardDutyBucket] PASSED                                    [ 71%]
tests/test_aws.py::test_metadata_version_services[AWSInspector] PASSED                                         [ 85%]
tests/test_aws.py::test_cloudtrail_db_maintenance[AWSCloudTrailBucket] PASSED                                  [100%]
```

**Comment by UranusBytes** [comment]
Hi @druizz90,

After I replaced the v3.8.2 of the aws-s3.py script with this version, on the first execution of the wodle, it didn't start from the last processed key/log.  Instead, it started from the value of `only_logs_after` configured for the bucket.  It only did this on the first execution.

A quick review of the code and SQL doesn't give me any reason to understand why this happened.  I didn't run in debug mode, so I don't have any exhaustive logs to know why this happened.  I'm unfortunately busy with other issues that prevent me from pulling a backup of my sqlite db to try to reproduce the error.

I would just suggest to take an existing v3.8.2 environment that has been polling logs (so it has some extensive history in sqlite db) and then upgrade the aws-s3.py script.  If that doesn't reprocess some of the logs, then there must have been some circumstance unique to my environment.

FWIW - my wodle config is like below, but with a large number of buckets

```
<wodle name="aws-s3">
  <disabled>no</disabled>
  <interval>10m</interval>
  <run_on_start>yes</run_on_start>
  <remove_from_bucket>no</remove_from_bucket>
  <bucket type="cloudtrail">
    <aws_account_alias>myAccount</aws_account_alias>
    <name>myBucketName</name>
    <aws_account_id>123412341234</aws_account_id>
    <iam_role_arn>arn:aws:iam::567856785678:role/myRoleName</iam_role_arn>
    <only_logs_after>2019-MAR-31</only_logs_after>
  </bucket>
</wodle>
```

Jeremy

> Hi @UranusBytes,
> 
> Could you give me more details about it? I can continue processing logs after last log which was processed properly.
> 
> Best regards,
> 
> Demetrio.

**Comment by druizz90** [comment]
> Hi @druizz90,
> 
> After I replaced the v3.8.2 of the aws-s3.py script with this version, on the first execution of the wodle, it didn't start from the last processed key/log. Instead, it started from the value of `only_logs_after` configured for the bucket. It only did this on the first execution.
> 
> A quick review of the code and SQL doesn't give me any reason to understand why this happened. I didn't run in debug mode, so I don't have any exhaustive logs to know why this happened. I'm unfortunately busy with other issues that prevent me from pulling a backup of my sqlite db to try to reproduce the error.
> 
> I would just suggest to take an existing v3.8.2 environment that has been polling logs (so it has some extensive history in sqlite db) and then upgrade the aws-s3.py script. If that doesn't reprocess some of the logs, then there must have been some circumstance unique to my environment.
> 
> FWIW - my wodle config is like below, but with a large number of buckets
> 
> ```
> <wodle name="aws-s3">
>   <disabled>no</disabled>
>   <interval>10m</interval>
>   <run_on_start>yes</run_on_start>
>   <remove_from_bucket>no</remove_from_bucket>
>   <bucket type="cloudtrail">
>     <aws_account_alias>myAccount</aws_account_alias>
>     <name>myBucketName</name>
>     <aws_account_id>123412341234</aws_account_id>
>     <iam_role_arn>arn:aws:iam::567856785678:role/myRoleName</iam_role_arn>
>     <only_logs_after>2019-MAR-31</only_logs_after>
>   </bucket>
> </wodle>
> ```
> 
> Jeremy
> 
> > Hi @UranusBytes,
> > Could you give me more details about it? I can continue processing logs after last log which was processed properly.
> > Best regards,
> > Demetrio.

Hi Jeremy,

Expected behaviour of `AWS` wodle is getting the value of `only_logs_after` only when DB is empty. I upgraded the wodle as you did and the wodle continues from the mark placed in DB, but it could be something that causes this error in a particular circumstance.

Thank you for your collaboration. Best regards,

Demetrio.

**Pull request merged**